### DISCUSSION TITLE
How to annotate a variable referring to an already defined function?

### DISCUSSION THREAD

**Discussion by geeshta** [python/typing#1038] [Q&A] [answered]
I ran into this while using the [xmlschema](https://github.com/sissaschool/xmlschema) package, so I will use it as an example. However the question could concern any `py.typed` library using variables as references to function.  
One of the classes have a `decode` method with the following annotation:
```python
def decode(
    self,
    source: Union[XMLSourceType, XMLResource],
    path: Optional[str] = None,
    schema_path: Optional[str] = None,
    validation: str = "strict",
    *args: Any,
    **kwargs: Any
) -> DecodeType[Any]: ...
```
this function is also reference by another variable creating an alias or a pointer to that function:
```python
to_dict = decode
```
The package is a `py.typed` package and such should follow [this documentation](https://github.com/python/typing/blob/master/docs/source/libraries.rst) which states, that "All variables have type annotations that refer to known types".
This also means that the `to_dict` variable should have annotations. If it doesn't, Pyright considers it `Unknown` and I get an error when calling `to_dict(...)` while using Pylance in VSCode. Also, because of this, code completion in the function call is lost. This only happens when importing from the library, when I open the source, the type of the variable is resolved correctly (locally).  And only happens with `py.typed` packages.

I was looking into `Callable` or even [PEP 677](https://www.python.org/dev/peps/pep-0677/) but neither of these support default arguments. So even if I edit the source file in the library to something like this:
```python
DecodeFuncType = Callable[
    [Union[XMLSourceType, XMLResource], str | None, str | None, str, Any, Any],
    DecodeType[Any],
]
to_dict: DecodeFuncType = decode
```
which follows the original function signature, I get limited completion, cannot use keyword arguments and typechecker complains about missing positional arguments. Also this is much more clunky when the function is already defined using the standard function annotation without Callable.

- How to correctly annotate such a variable, that refers to a function defined with standard inline annotations?
- Is there a way to annotate them, such that keyword parameters and default values are included?
- Shouldn't such a case be an exception in the documentation section "Type annotations can be omitted in a few specific cases where the type is obvious from the context"

Steps to reproduce:

1. Use VSCode with Pylance and strict type checking in the extension settings
2. Install the `xmlschema` package
3. Enter the following code
 ```python
import xmlschema


xsd = xmlschema.XMLSchema("some_schema.xsd")

xml_dict = xsd.to_dict("some_xml.xml") # Pylance complains that `to_dict()` is Unknown and no code completion is provided
xml_decoded = xsd.decode("some_xml.xml") # This is ok and code completion is provided
```

**Comment by TeamSpen210** [comment, marked as answer]
There’s two ways you can annotate this. First you could just duplicate the function definition under the second name. Secondly, if you do want to specify it as more of a global variable that may be changed, you can use a `Protocol` containing only a `__call__` method as effectively an advanced `Callable` type.
